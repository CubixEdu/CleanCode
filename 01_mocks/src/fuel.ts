export class Fuel {

    public fill(litersOfFuel: number) {
        return litersOfFuel;
    }
}