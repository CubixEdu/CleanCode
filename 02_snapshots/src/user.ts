export class User {
  
    constructor(
      private id: number,
      private username: string,
      private email: string,
      private firstName: string,
      private lastName: string,
      private dateOfBirth: Date,
      private address: {
        street: string;
        city: string;
        zipCode: string;
      },
      private phoneNumber: string,
      private isActive: boolean,
      private roles: string[]
    ) {}
  }