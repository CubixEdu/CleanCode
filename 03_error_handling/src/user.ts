export class User {
  
    constructor(
      private id: number,
      private username: string,
      private email: string,
      private firstName: string,
      private lastName: string,
      private dateOfBirth: Date
    ) {}
  }