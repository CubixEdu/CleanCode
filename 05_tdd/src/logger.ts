export class Logger {

    public logError(errorMessage: string) {
        console.log(errorMessage);
    }
}