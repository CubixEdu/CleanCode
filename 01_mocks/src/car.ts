import { Fuel } from "./fuel";

export class Car {

    constructor(private fuel: Fuel) { }

    public drive() {
        this.fuel.fill(35);
    }
}