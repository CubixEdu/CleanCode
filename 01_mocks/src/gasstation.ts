import { Fuel } from "./fuel";

export class GasStation {

    constructor(private fuel: Fuel) { }

    public fillUpCars(numberOfCars: number) {
        for (let index = 1; index < numberOfCars + 1; index++) {
            this.fuel.fill(index + 1);
        }
    }
}