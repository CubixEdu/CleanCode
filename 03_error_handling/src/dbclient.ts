import { User } from "./user";

export class DbClient {

    public getUserByUsername(username: string) {
        return new User(
            1,
            'john_doe',
            'john@example.com',
            'John',
            'Doe',
            new Date('1990-01-15')
          );
    }
}