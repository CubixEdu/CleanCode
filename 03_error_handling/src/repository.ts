import { NetworkError } from "./exceptions/NetworkError";
import { NotFoundError } from "./exceptions/NotFoundError";
import { UnknownError } from "./exceptions/UnknownError";
import { ValidationError } from "./exceptions/ValidationError";
import { DbClient } from "./dbclient";
import { Logger } from "./logger";

export class Repository {

    constructor(private dbClient: DbClient, private logger: Logger) { }

    public getUserByUsername(username: string) {
        this.validateInput(username);

        try {
            return this.dbClient.getUserByUsername(username);
        } catch (error) {
            if (error instanceof NetworkError || error instanceof NotFoundError) {
                this.logger.logError(error.message, error);
                throw error;
            }

            throw new UnknownError('Unknown error happened.');
        }
    }

    private validateInput(username: string) {
        if (!username) {
            throw new ValidationError('username cannot be null or empty.');
        }
    }
}