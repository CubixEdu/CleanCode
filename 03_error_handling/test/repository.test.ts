import { DbClient } from "../src/dbclient";
import { Logger } from "../src/logger";
import { Repository } from "../src/repository"
import { mock, mockReset } from 'jest-mock-extended';
import { User } from "../src/user";
import { NetworkError } from "../src/exceptions/NetworkError";
import { NotFoundError } from "../src/exceptions/NotFoundError";
import { ValidationError } from "../src/exceptions/ValidationError";

describe('Repository tests', () => {
    let sut: Repository;
    const mockDbClient = mock<DbClient>();
    const mockLogger = mock<Logger>();
    const username = 'batman';

    beforeEach(() => {
        mockReset(mockDbClient);
        mockReset(mockLogger);

        sut = new Repository(mockDbClient, mockLogger);
    })

    describe('Happy path', () => {
        it('should return the user when retrieved by username', () => {
            // Arrange
            const user = new User(
                1,
                'john_doe',
                'john@example.com',
                'John',
                'Doe',
                new Date('1990-01-15')
            );
            mockDbClient.getUserByUsername.mockReturnValue(user);

            // Act
            const result = sut.getUserByUsername(username);

            // Assert
            expect(result).toMatchSnapshot();
            expect(mockDbClient.getUserByUsername).toBeCalledTimes(1);
            expect(mockDbClient.getUserByUsername).toBeCalledWith(username);
            expect(mockLogger.logError).toBeCalledTimes(0);
        })
    })

    describe('Error paths', () => {
        it('should throw a not found error if the user does not exist', () => {
            // Arrange
            const errorMessage = "User was not found.";
            const expectedError = new NotFoundError(errorMessage);
            mockDbClient.getUserByUsername.mockImplementation(() => { throw expectedError });

            // Act
            expect(() => sut.getUserByUsername(username)).toThrow(expectedError)
            expect(mockDbClient.getUserByUsername).toBeCalledTimes(1);
            expect(mockDbClient.getUserByUsername).toBeCalledWith(username)
            expect(mockLogger.logError).toBeCalledTimes(1);
            expect(mockLogger.logError).toBeCalledWith(errorMessage, expectedError);
        })

        it('should throw a network error if the DatabaseClient cannot connect to the database', () => {
            // Arrange
            const errorMessage = "Network failure happened.";
            const expectedError = new NetworkError(errorMessage);
            mockDbClient.getUserByUsername.mockImplementation(() => { throw expectedError });

            // Act
            expect(() => sut.getUserByUsername(username)).toThrow(expectedError)
            expect(mockDbClient.getUserByUsername).toBeCalledTimes(1);
            expect(mockDbClient.getUserByUsername).toBeCalledWith(username)
            expect(mockLogger.logError).toBeCalledTimes(1);
            expect(mockLogger.logError).toBeCalledWith(errorMessage, expectedError);
        })

        it('should throw a validation error if the username is empty', () => {
            // Arrange
            const errorMessage = 'username cannot be null or empty.';
            const expectedError = new ValidationError(errorMessage);
            mockDbClient.getUserByUsername.mockImplementation(() => { throw expectedError });

            // Act
            expect(() => sut.getUserByUsername('')).toThrow(expectedError)
            expect(mockDbClient.getUserByUsername).toBeCalledTimes(0);
            expect(mockLogger.logError).toBeCalledTimes(0);
        })
    })

})