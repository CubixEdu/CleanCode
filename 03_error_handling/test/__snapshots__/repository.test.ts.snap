// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository tests Happy path should return the user when retrieved by username 1`] = `
User {
  "dateOfBirth": 1990-01-15T00:00:00.000Z,
  "email": "john@example.com",
  "firstName": "John",
  "id": 1,
  "lastName": "Doe",
  "username": "john_doe",
}
`;
